### PR TITLE
Alphanumeric notebook entry

### PR DESCRIPTION
**Pull request by nikhilmandlik**
### Resolves: 
[Notebook] Copy label and value from alphanumeric in Layout directly to a Notebook entry #3299 

### TODOs:
- merge branch [three-dot-menu-proto](https://github.com/nasa/openmct/tree/three-dot-menu-proto)

### Notes: 
- based of branch [three-dot-menu-proto](https://github.com/nasa/openmct/tree/three-dot-menu-proto)

### Author Checklist:
Check | Passed
-- | --
Changes address original issue? | Y
Unit tests included and/or updated with changes? | N/A
Command line build passes? | Y
Changes have been smoke-tested? | Y
Testing instructions included? | Y

### Screenshots:
Screen Shot 2020-09-21 at 12 45 39 PM

### CODE REVIEW

File: src/plugins/displayLayout/AlphanumericFormatViewProvider.js [hunk 1]
```diff
@@ -64,9 +64,17 @@ define([
                             components: {
                                 AlphanumericFormatView: AlphanumericFormatView.default
                             },
-                            template: '<alphanumeric-format-view></alphanumeric-format-view>'
+                            template: '<alphanumeric-format-view ref="alphanumericFormatView"></alphanumeric-format-view>'
                         });
                     },
+                    getViewContext() {
+                        console.log('getViewContext', component);
```
Comment (deeptailor) [inline src/plugins/displayLayout/AlphanumericFormatViewProvider.js:71]: remove the console.log

File: src/plugins/displayLayout/actions/CopyToClipboardAction.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        this.cssClass = 'icon-duplicate';
+        this.description = 'Copy to Clipboard action';
+        this.group = "action";
+        this.key = 'copyToClipboard';
+        this.name = 'Copy to Clipboard';
+        this.priority = 9;
+    }
+
+    invoke(objectPath = null, viewContext) {
+        if (!objectPath) {
```
Comment (deeptailor) [inline src/plugins/displayLayout/actions/CopyToClipboardAction.js:16]: why are we returning if we dont have an object path? we dont need an objectPath for this to work right?

File: src/plugins/displayLayout/components/TelemetryView.vue [hunk 3]
```diff
@@ -284,14 +298,31 @@ export default {
...
+            const defaultNotebook = getDefaultNotebook();
+
+            const domainObject = defaultNotebook && await this.openmct.objects.get(defaultNotebook.notebookMeta.identifier);
+
+            const actionsObject = this.openmct.actions.get(this.currentObjectPath, this.getViewContext(), { viewHistoricalData: true }).applicableActions;
+            let applicableActionKeys = Object.keys(actionsObject)
+                .filter(key => {
+                    const isViewHistoricalData = actionsObject[key].key === 'viewHistoricalData';
+                    const isCopyToClipboard = actionsObject[key].key === 'copyToClipboard';
+                    const isCopyToNotebook = defaultNotebook && actionsObject[key].key === 'copyToNotebook';
```
Comment (deeptailor) [inline src/plugins/displayLayout/components/TelemetryView.vue:311]: lets move this to the constant as it was, so in the future, if we needed to add more actions to this menu, we can just update the array at the top

File: src/plugins/displayLayout/components/TelemetryView.vue [hunk 4]
```diff
@@ -284,14 +298,31 @@ export default {
...
+
+                    if (isCopyToNotebook) {
+                        const defaultPath = domainObject && `${domainObject.name} - ${defaultNotebook.section.name} - ${defaultNotebook.page.name}`;
+                        actionsObject[key].name = `Copy to Notebook ${defaultPath}`;
+                    }
+
+                    return isCopyToClipboard || isCopyToNotebook || isViewHistoricalData;
```
Comment (deeptailor) [inline src/plugins/displayLayout/components/TelemetryView.vue:318]: see comment above

File: src/plugins/displayLayout/components/TelemetryView.vue [hunk 5]
```diff
@@ -284,14 +298,31 @@ export default {
...
+                });
 
-            this.contextMenuActions = applicableActionKeys.map(key => actionsObject[key]);
+            return applicableActionKeys.map(key => actionsObject[key]);
         },
-        showContextMenu(event) {
-            this.openmct.menus.showMenu(event.x, event.y, this.contextMenuActions);
+        async showContextMenu(event) {
+            const contextMenuActions = await this.getContextMenuActions();
```
Comment (deeptailor) [inline src/plugins/displayLayout/components/TelemetryView.vue:324]: this should not be async

File: src/plugins/notebook/actions/CopyToNotebookAction.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+                const defaultPath = `${domainObject.name} - ${notebookStorage.section.name} - ${notebookStorage.page.name}`;
+                const msg = `Saved to Notebook ${defaultPath}`;
+                this.openmct.notifications.info(msg);
+            });
+    }
+
+    invoke(objectPath = null, viewContext) {
+        if (!objectPath) {
+            return;
```
Comment (deeptailor) [inline src/plugins/notebook/actions/CopyToNotebookAction.js:30]: see comment above regarding returning if objectPath is null

File: src/utils/clipboard.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+        this.clipboardRead = false;
+        this.clipboardWrite = false;
+
+        this.init();
+
+        Clipboard.instance = this;
+    }
+
+    // methods
```
Comment (deeptailor) [inline src/utils/clipboard.js:25]: no need for this comment, its pretty cleat that the following are methods

File: src/utils/clipboard.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    readClipboard() {
+        if (!this.clipboardRead) {
+            return Promise.reject('Error: No permission to read clipboard');
+        }
+
+        // return promise
+        return navigator.clipboard.readText();
+    }
+}
```
Comment (deeptailor) [inline src/utils/clipboard.js:59]: if we do const clipBoard = new Clipboard(), then return clipBoard in the getClipboard function, we dont need to check if there is an instance already. That way we can make sure its a singleton returned whenever its called.

**Review by deeptailor — CHANGES_REQUESTED** [review]
Looks great, just a few comments

File: src/utils/clipboard.js (hunk 8 above)
Comment (nikhilmandlik) [reply src/utils/clipboard.js:59]: yeah was doing that before, but changed due to https://eslint.org/docs/rules/no-constructor-return

File: src/plugins/displayLayout/components/TelemetryView.vue (hunk 5 above)
Comment (nikhilmandlik) [reply src/plugins/displayLayout/components/TelemetryView.vue:324]: Actually we need it as calling async getContextMenuActions inside this.

File: src/utils/clipboard.js (hunk 8 above)
Comment (deeptailor) [reply src/utils/clipboard.js:58]: According to the no-constructor-return rule, we should not be returning out of the constructor, and that is exactly what you are doing on line 14.

The comment I wrote earlier still stands. Let me know if it doesnt make sense.

**Review by deeptailor — CHANGES_REQUESTED** [review]
Looks good, still needs one change

File: src/utils/clipboard.js (hunk 8 above)
Comment (nikhilmandlik) [reply src/utils/clipboard.js:58]: Actually its `Disallow returning value` but it's ok to have `return`.
Lets have a call, I think I am missing the point you are trying to make.

**Review by deeptailor — APPROVED** [review]
LGTM

**Pull request merged**